src/model/test/AirLoopHVACUnitaryHeatPumpAirToAirMultiSpeed_GTest.cpp

```diff
@@ -10,5 +10,338 @@
+  EXPECT_EQ(3.1, hp.speed3SupplyAirFlowRateDuringCoolingOperation().get());
+  // Bad Value
+  EXPECT_FALSE(hp.setSpeed3SupplyAirFlowRateDuringCoolingOperation(-10.0));
+  ASSERT_TRUE(hp.speed3SupplyAirFlowRateDuringCoolingOperation());
+  EXPECT_EQ(3.1, hp.speed3SupplyAirFlowRateDuringCoolingOperation().get());
+  EXPECT_FALSE(hp.isSpeed3SupplyAirFlowRateDuringCoolingOperationAutosized());
+
+  // Speed 4 Supply Air Flow Rate During Cooling Operation: Required Double
+  // Autosize
+  hp.autosizeSpeed4SupplyAirFlowRateDuringCoolingOperation();
+  EXPECT_TRUE(hp.isSpeed4SupplyAirFlowRateDuringCoolingOperationAutosized());
+  // Set
+  EXPECT_TRUE(hp.setSpeed4SupplyAirFlowRateDuringCoolingOperation(3.2));
+  ASSERT_TRUE(hp.speed4SupplyAirFlowRateDuringCoolingOperation());
+  EXPECT_EQ(3.2, hp.speed4SupplyAirFlowRateDuringCoolingOperation().get());
+  // Bad Value
+  EXPECT_FALSE(hp.setSpeed4SupplyAirFlowRateDuringCoolingOperation(-10.0));
+  ASSERT_TRUE(hp.speed4SupplyAirFlowRateDuringCoolingOperation());
+  EXPECT_EQ(3.2, hp.speed4SupplyAirFlowRateDuringCoolingOperation().get());
+  EXPECT_FALSE(hp.isSpeed4SupplyAirFlowRateDuringCoolingOperationAutosized());
+}
+
+TEST_F(ModelFixture, AirLoopHVACUnitaryHeatPumpAirToAirMultiSpeed_addToNode) {
+  Model m;
+  Schedule s = m.alwaysOnDiscreteSchedule();
+  FanConstantVolume supplyFan(m, s);
+
+  CoilCoolingDXMultiSpeed cc(m);
+  CoilHeatingElectricMultiStage hc(m);
+
+  CoilHeatingElectric suppHC(m);
+
+  AirLoopHVACUnitaryHeatPumpAirToAirMultiSpeed testObject(m, supplyFan, hc, cc, suppHC);
+
+  AirLoopHVAC airLoop(m);
+
+  Node supplyOutletNode = airLoop.supplyOutletNode();
+  EXPECT_EQ(2, airLoop.supplyComponents().size());
+  EXPECT_TRUE(testObject.addToNode(supplyOutletNode));
+  EXPECT_EQ(3, airLoop.supplyComponents().size());
+
+  EXPECT_EQ(5, airLoop.demandComponents().size());
+  Node inletNode = airLoop.zoneSplitter().lastOutletModelObject()->cast<Node>();
+  EXPECT_FALSE(testObject.addToNode(inletNode));
+  EXPECT_EQ(5, airLoop.demandComponents().size());
+
+  PlantLoop plantLoop(m);
+  supplyOutletNode = plantLoop.supplyOutletNode();
+  EXPECT_EQ(5, plantLoop.supplyComponents().size());
+  EXPECT_FALSE(testObject.addToNode(supplyOutletNode));
+  EXPECT_EQ(5, plantLoop.supplyComponents().size());
+
+  Node demandOutletNode = plantLoop.demandOutletNode();
+  EXPECT_EQ(5, plantLoop.demandComponents().size());
+  EXPECT_FALSE(testObject.addToNode(demandOutletNode));
+  EXPECT_EQ(5, plantLoop.demandComponents().size());
+
+  auto testObjectClone = testObject.clone(m).cast<AirLoopHVACUnitaryHeatPumpAirToAirMultiSpeed>();
+  // Clone should reset the inlet/outlet nodes
+  ASSERT_FALSE(testObjectClone.inletModelObject());
+  ASSERT_FALSE(testObjectClone.outletModelObject());
+
+  supplyOutletNode = airLoop.supplyOutletNode();
+  EXPECT_TRUE(testObjectClone.addToNode(supplyOutletNode));
+  EXPECT_EQ(5, airLoop.supplyComponents().size());
+}
```

Comment:
Completely wrong originally.

The Clone method was using MdoelObject_Impl::clone, which does not reset the inlet/outlet nodes.
And the addToNode was not overriden so it'd accept ANY connection.


```
[ RUN      ] ModelFixture.AirLoopHVACUnitaryHeatPumpAirToAirMultiSpeed_addToNode
/media/DataExt4/Software/Others/OpenStudio2/src/model/test/AirLoopHVACUnitaryHeatPumpAirToAirMultiSpeed_GTest.cpp:310: Failure
Value of: testObject.addToNode(inletNode)
  Actual: true
Expected: false

/media/DataExt4/Software/Others/OpenStudio2/src/model/test/AirLoopHVACUnitaryHeatPumpAirToAirMultiSpeed_GTest.cpp:311: Failure
Expected equality of these values:
  5
  airLoop.demandComponents().size()
    Which is: 7

/media/DataExt4/Software/Others/OpenStudio2/src/model/test/AirLoopHVACUnitaryHeatPumpAirToAirMultiSpeed_GTest.cpp:316: Failure
Value of: testObject.addToNode(supplyOutletNode)
  Actual: true
Expected: false

/media/DataExt4/Software/Others/OpenStudio2/src/model/test/AirLoopHVACUnitaryHeatPumpAirToAirMultiSpeed_GTest.cpp:317: Failure
Expected equality of these values:
  5
  plantLoop.supplyComponents().size()
    Which is: 7

/media/DataExt4/Software/Others/OpenStudio2/src/model/test/AirLoopHVACUnitaryHeatPumpAirToAirMultiSpeed_GTest.cpp:321: Failure
Value of: testObject.addToNode(demandOutletNode)
  Actual: true
Expected: false

/media/DataExt4/Software/Others/OpenStudio2/src/model/test/AirLoopHVACUnitaryHeatPumpAirToAirMultiSpeed_GTest.cpp:322: Failure
Expected equality of these values:
  5
  plantLoop.demandComponents().size()
    Which is: 7

/media/DataExt4/Software/Others/OpenStudio2/src/model/test/AirLoopHVACUnitaryHeatPumpAirToAirMultiSpeed_GTest.cpp:331: Failure
Expected equality of these values:
  5
  airLoop.supplyComponents().size()
    Which is: 3
```